src/base/checked.ml

```diff
@@ -72,5 +76,12 @@ module Make
         bind acc ~f:(fun () ->
             add_constraint (Constraint.override_label c label) ) )
 
-  let assert_equal ?label x y = assert_ (Constraint.equal ?label x y)
+  let assert_equal ?label x y =
+    match (x, y) with
+    | Cvar.Constant x, Cvar.Constant y ->
+        if Field.equal x y then return ()
+        else
+          failwithf !"assert_equal: %{sexp: Field.t} != %{sexp: Field.t}" x y ()
+    | _ ->
+        assert_ (Constraint.equal ?label x y)
```

Comment:
if you're wondering where this leads to, here's the code from plonk_constraint_system:

```ocaml
    | Snarky_backendless.Constraint.Equal (v1, v2) -> (
        (* `red` will create rows in the circuit to encode the variables if need be *)
        let (s1, x1), (s2, x2) = (red v1, red v2) in
        match (x1, x2) with
        | `Var x1, `Var x2 ->
            (* if the scalars match, we can optimize the equality check *)
            if Fp.equal s1 s2 then (
              (* if the scalars are 0, we do nothing *)
              if not (Fp.equal s1 Fp.zero) then
                (* we optimize the equality check by just using the permutation, see https://o1-labs.github.io/proof-systems/snarky/kimchi-backend.html#union-finds *)
                Union_find.union (union_find sys x1) (union_find sys x2) )
            else if (* s1 x1 - s2 x2 = 0
          *)
                    not (Fp.equal s1 s2) then
              add_generic_constraint ~l:x1 ~r:x2
                [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
                sys
            else
              add_generic_constraint ~l:x1 ~r:x2
                [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
                sys
```